cranelift/codegen/build.rs

```diff
@@ -259,3 +342,70 @@ fn rebuild_isle(crate_dir: &std::path::Path) -> Result<(), Box<dyn std::error::E
     }
 }
+
+/// Check that the generated source, which we are *not* rebuilding in
+/// this configuration (the default) and instead relying on the
+/// checked-into-git generated copy, still corresponds to the ISLE
+/// source. We do this by computing a hash of the ISLE source and
+/// checking it against a "manifest" that is also checked into git,
+/// alongside the generated Rust. If there is a mismatch, we signal a
+/// build error and suggest to the user that they might wish to build
+/// with the `rebuild-isle` feature.
+///
+/// (Why not include this feature by default? Because the build
+/// process must not modify the checked-in source by default; any
+/// checked-in source is a human-managed bit of data, and we can only
+/// act as an agent of the human developer when explicitly requested
+/// to do so. This manifest check is a middle ground that ensures this
+/// explicit control while also avoiding the easy footgun of "I
+/// changed the ISLE, why isn't the compiler updated?!".)
+#[cfg(not(feature = "rebuild-isle"))]
+fn check_isle_manifests(
+    crate_dir: &std::path::Path,
+) -> Result<(), Box<dyn std::error::Error + 'static>> {
+    let isle_compilations = get_isle_compilations(crate_dir)?;
+
+    let mut had_error = false;
+    for (out_file, files) in &isle_compilations {
+        for file in files {
+            println!("cargo:rerun-if-changed={}", file.display());
+        }
+
+        let manifest = std::fs::read_to_string(manifest_filename(out_file))?;
+        let expected_manifest = compute_isle_source_manifest(&files[..])?;
+        if manifest != expected_manifest {
+            had_error = true;
+            eprintln!("");
+            eprintln!("Error: the ISLE source files that resulted in the generated Rust source");
+            eprintln!("");
+            eprintln!("      * {}", out_file.display());
+            eprintln!("");
+            eprintln!(
+                "have changed but the generated source was not rebuilt! These ISLE source"
+            );
+            eprintln!("files are:");
+            eprintln!("");
+            for file in files {
+                eprintln!("       * {}", file.display());
+            }
+        }
+    }
+
+    if had_error {
+        eprintln!("");
+        eprintln!("Please add `--features rebuild-isle` to your `cargo build` command");
+        eprintln!("if you wish to rebuild the generated source, then include these changes");
+        eprintln!("in any git commits you make that include the changes to the ISLE.");
+        eprintln!("");
+        eprintln!("For example:");
+        eprintln!("");
+        eprintln!("  $ cargo build -p cranelift-codegen --features rebuild-isle");
+        eprintln!("");
+        eprintln!("(This build script cannot do this for you by default because we cannot");
+        eprintln!("modify checked-into-git source without your explicit opt-in.)");
+        eprintln!("");
+        std::process::abort();
```

Comment:
FWIW I see that this was done already but `std::process::abort` is a somewhat violent method of taking down the process (I think it may core dump?) and `std::process::exit` I think is a reasonable alternative.

Reply:
Yep, I was following the existing `abort()` but I've made it `exit(1)` now.